File: app_pojav_zh/src/main/java/com/movtery/pojavzh/ui/fragment/ProfilePathManagerFragment.java
package com.movtery.pojavzh.ui.fragment;

import android.os.Bundle;
import android.os.Environment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;
import android.view.animation.LayoutAnimationController;
import android.widget.ImageButton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.daimajia.androidanimations.library.Techniques;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.movtery.pojavzh.feature.customprofilepath.ProfilePathJsonObject;
import com.movtery.pojavzh.feature.customprofilepath.ProfilePathManager;
import com.movtery.pojavzh.ui.dialog.EditTextDialog;
import com.movtery.pojavzh.ui.subassembly.customprofilepath.ProfileItem;
import com.movtery.pojavzh.ui.subassembly.customprofilepath.ProfilePathAdapter;
import com.movtery.pojavzh.utils.ZHTools;
import com.movtery.pojavzh.utils.anim.OnSlideOutListener;
import com.movtery.pojavzh.utils.anim.ViewAnimUtils;

import net.kdt.pojavlaunch.R;
import net.kdt.pojavlaunch.Tools;
import net.kdt.pojavlaunch.extra.ExtraConstants;
import net.kdt.pojavlaunch.extra.ExtraCore;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

public class ProfilePathManagerFragment extends FragmentWithAnim {
    public static final String TAG = "ProfilePathManagerFragment";
    private final List<ProfileItem> mData = new ArrayList<>();
    private View mPathLayout, mOperateLayout, mShadowView, mOperateView;
    private ProfilePathAdapter adapter;

    public ProfilePathManagerFragment() {
        super(R.layout.fragment_profile_path_manager);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        String value = (String) ExtraCore.consumeValue(ExtraConstants.FILE_SELECTOR);

        if (value != null && !value.isEmpty() && !isAddedPath(value)) {
            new EditTextDialog.Builder(requireContext())
                    .setTitle(R.string.zh_profiles_path_create_new_title)
                    .setConfirmListener(editBox -> {
                        String string = editBox.getText().toString();
                        if (string.isEmpty()) {
                            editBox.setError(getString(R.string.global_error_field_empty));
                            return false;
                        }

                        this.mData.add(new ProfileItem(UUID.randomUUID().toString(), string, value));
                        ProfilePathManager.save(this.mData);
                        refresh();

                        return true;
                    }).buildDialog();
        }

        return super.onCreateView(inflater, container, savedInstanceState);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        refreshData();

        mPathLayout = view.findViewById(R.id.path_layout);
        mOperateLayout = view.findViewById(R.id.operate_layout);
        mShadowView = view.findViewById(R.id.shadowView);

        mOperateView = view.findViewById(R.id.operate_view);

        RecyclerView pathList = view.findViewById(R.id.zh_profile_path);
        ImageButton refreshButton = view.findViewById(R.id.zh_profile_path_refresh_button);
        ImageButton createNewButton = view.findViewById(R.id.zh_profile_path_create_new_button);
        ImageButton returnButton = view.findViewById(R.id.zh_profile_path_return_button);

        ZHTools.setTooltipText(refreshButton, refreshButton.getContentDescription());
        ZHTools.setTooltipText(createNewButton, createNewButton.getContentDescription());
        ZHTools.setTooltipText(returnButton, returnButton.getContentDescription());

        adapter = new ProfilePathAdapter(this, pathList, this.mData);
        pathList.setLayoutAnimation(new LayoutAnimationController(AnimationUtils.loadAnimation(view.getContext(), R.anim.fade_downwards)));
        pathList.setLayoutManager(new LinearLayoutManager(requireContext()));
        pathList.setAdapter(adapter);

        refreshButton.setOnClickListener(v -> refresh());
        createNewButton.setOnClickListener(v -> {
            Bundle bundle = new Bundle();
            bundle.putBoolean(FilesFragment.BUNDLE_SELECT_FOLDER_MODE, true);
            bundle.putBoolean(FilesFragment.BUNDLE_SHOW_FILE, false);
            bundle.putBoolean(FilesFragment.BUNDLE_REMOVE_LOCK_PATH, false);
            bundle.putString(FilesFragment.BUNDLE_LOCK_PATH, Environment.getExternalStorageDirectory().getAbsolutePath());

            ZHTools.swapFragmentWithAnim(this, FilesFragment.class, FilesFragment.TAG, bundle);
        });
        returnButton.setOnClickListener(v -> ZHTools.onBackPressed(requireActivity()));

        ViewAnimUtils.slideInAnim(this);
    }

    private void refresh() {
        refreshData();
        adapter.updateData(this.mData);
    }

    private void refreshData() {
        this.mData.clear();
        this.mData.add(new ProfileItem("default", getString(R.string.zh_profiles_path_default), Tools.DIR_GAME_HOME));

        try {
            String json;
            if (ZHTools.FILE_PROFILE_PATH.exists()) {
                json = Tools.read(ZHTools.FILE_PROFILE_PATH);
                if (json.isEmpty()) {
                    return;
                }
            } else {
                return;
            }

            JsonObject jsonObject = JsonParser.parseString(json).getAsJsonObject();

            for (String key : jsonObject.keySet()) {
                ProfilePathJsonObject profilePathId = new Gson().fromJson(jsonObject.get(key), ProfilePathJsonObject.class);
                ProfileItem item = new ProfileItem(key, profilePathId.title, profilePathId.path);
                this.mData.add(item);
            }
        } catch (Exception ignored) {
        }
    }

    private boolean isAddedPath(String path) {
        for (ProfileItem mDatum : this.mData) {
            if (Objects.equals(mDatum.path, path)) {
                return true;
            }
        }
        return false;
    }

    @Override
    public void slideIn() {
        ViewAnimUtils.setViewAnim(mPathLayout, Techniques.BounceInDown);
        ViewAnimUtils.setViewAnim(mOperateLayout, Techniques.BounceInLeft);
        ViewAnimUtils.setViewAnim(mShadowView, Techniques.BounceInLeft);
        ViewAnimUtils.setViewAnim(mOperateView, Techniques.FadeInLeft);
    }

    @Override
    public void slideOut(@NonNull OnSlideOutListener listener) {
        ViewAnimUtils.setViewAnim(mPathLayout, Techniques.FadeOutUp);
        ViewAnimUtils.setViewAnim(mOperateLayout, Techniques.FadeOutRight);
        ViewAnimUtils.setViewAnim(mShadowView, Techniques.FadeOutRight);
        super.slideOut(listener);
    }
}

File: app_pojav_zh/src/main/java/com/movtery/pojavzh/ui/fragment/ProfilePathManagerFragment.kt
package com.movtery.pojavzh.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import android.view.animation.LayoutAnimationController
import android.widget.EditText
import androidx.recyclerview.widget.LinearLayoutManager
import com.getkeepsafe.taptargetview.TapTargetSequence
import com.google.gson.Gson
import com.google.gson.JsonParser
import com.movtery.anim.AnimPlayer
import com.movtery.anim.animations.Animations
import com.movtery.pojavzh.event.sticky.FileSelectorEvent
import com.movtery.pojavzh.feature.customprofilepath.ProfilePathJsonObject
import com.movtery.pojavzh.feature.customprofilepath.ProfilePathManager.Companion.save
import com.movtery.pojavzh.feature.log.Logging
import com.movtery.pojavzh.ui.dialog.EditTextDialog
import com.movtery.pojavzh.ui.subassembly.customprofilepath.ProfileItem
import com.movtery.pojavzh.ui.subassembly.customprofilepath.ProfilePathAdapter
import com.movtery.pojavzh.utils.NewbieGuideUtils
import com.movtery.pojavzh.utils.PathAndUrlManager
import com.movtery.pojavzh.utils.ZHTools
import net.kdt.pojavlaunch.R
import net.kdt.pojavlaunch.Tools
import net.kdt.pojavlaunch.databinding.FragmentProfilePathManagerBinding
import org.greenrobot.eventbus.EventBus
import java.util.UUID

class ProfilePathManagerFragment : FragmentWithAnim(R.layout.fragment_profile_path_manager) {
    companion object {
        const val TAG: String = "ProfilePathManagerFragment"
    }

    private lateinit var binding: FragmentProfilePathManagerBinding
    private val mData: MutableList<ProfileItem> = ArrayList()
    private var profilePathAdapter: ProfilePathAdapter? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val value = EventBus.getDefault().getStickyEvent(FileSelectorEvent::class.java)?.path

        value?.let {
            if (value.isNotEmpty() && !isAddedPath(value)) {
                EditTextDialog.Builder(requireContext())
                    .setTitle(R.string.profiles_path_create_new_title)
                    .setConfirmListener { editBox: EditText ->
                        val string = editBox.text.toString()
                        if (string.isEmpty()) {
                            editBox.error = getString(R.string.generic_error_field_empty)
                            return@setConfirmListener false
                        }

                        mData.add(ProfileItem(UUID.randomUUID().toString(), string, value))
                        save(this.mData)
                        refresh()
                        true
                    }.buildDialog()
            }
        }

        binding = FragmentProfilePathManagerBinding.inflate(layoutInflater)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        refreshData()

        binding.apply {
            profilePathAdapter = ProfilePathAdapter(this@ProfilePathManagerFragment, recyclerView, mData)
            recyclerView.apply {
                layoutAnimation = LayoutAnimationController(
                    AnimationUtils.loadAnimation(view.context, R.anim.fade_downwards)
                )
                layoutManager = LinearLayoutManager(requireContext())
                this.adapter = profilePathAdapter
            }

            refreshButton.setOnClickListener { refresh() }
            createNewButton.setOnClickListener {
                val bundle = Bundle()
                bundle.putBoolean(FilesFragment.BUNDLE_SELECT_FOLDER_MODE, true)
                bundle.putBoolean(FilesFragment.BUNDLE_SHOW_FILE, false)
                bundle.putBoolean(FilesFragment.BUNDLE_REMOVE_LOCK_PATH, false)
                ZHTools.swapFragmentWithAnim(this@ProfilePathManagerFragment, FilesFragment::class.java, FilesFragment.TAG, bundle)
            }
            returnButton.setOnClickListener { ZHTools.onBackPressed(requireActivity()) }

            ZHTools.setTooltipText(
                refreshButton,
                createNewButton,
                returnButton
            )

            if (NewbieGuideUtils.showOnlyOne(TAG)) return
            val fragmentActivity = requireActivity()
            TapTargetSequence(fragmentActivity)
                .targets(
                    NewbieGuideUtils.getSimpleTarget(fragmentActivity, refreshButton, getString(R.string.generic_refresh), getString(R.string.newbie_guide_general_refresh)),
                    NewbieGuideUtils.getSimpleTarget(fragmentActivity, createNewButton, getString(R.string.profiles_path_create_new), getString(R.string.newbie_guide_profiles_path_create)),
                    NewbieGuideUtils.getSimpleTarget(fragmentActivity, returnButton, getString(R.string.generic_close), getString(R.string.newbie_guide_general_close)))
                .start()
        }
    }

    private fun refresh() {
        refreshData()
        profilePathAdapter?.updateData(this.mData)
    }

    private fun refreshData() {
        mData.clear()
        mData.add(ProfileItem("default", getString(R.string.profiles_path_default), PathAndUrlManager.DIR_GAME_HOME))

        runCatching {
            val json: String
            if (PathAndUrlManager.FILE_PROFILE_PATH.exists()) {
                json = Tools.read(PathAndUrlManager.FILE_PROFILE_PATH)
                if (json.isEmpty()) return
            } else return

            val jsonObject = JsonParser.parseString(json).asJsonObject

            for (key in jsonObject.keySet()) {
                val profilePathId = Gson().fromJson(jsonObject[key], ProfilePathJsonObject::class.java)
                val item = ProfileItem(key, profilePathId.title, profilePathId.path)
                mData.add(item)
            }
        }.getOrElse { e -> Logging.e("refresh profile data", Tools.printToString(e)) }
    }

    private fun isAddedPath(path: String): Boolean {
        for (mDatum in this.mData) {
            if (mDatum.path == path) return true
        }
        return false
    }

    override fun slideIn(animPlayer: AnimPlayer) {
        binding.apply {
            animPlayer.apply(AnimPlayer.Entry(pathLayout, Animations.BounceInDown))
                .apply(AnimPlayer.Entry(operateLayout, Animations.BounceInLeft))
                .apply(AnimPlayer.Entry(operateView, Animations.FadeInLeft))
        }
    }

    override fun slideOut(animPlayer: AnimPlayer) {
        binding.apply {
            animPlayer.apply(AnimPlayer.Entry(pathLayout, Animations.FadeOutUp))
                .apply(AnimPlayer.Entry(operateLayout, Animations.FadeOutRight))
        }
    }
}
